### DISCUSSION TITLE
Scope selection rules in test_plan.py script and twisters workflows

### DISCUSSION THREAD

**Discussion by PerMac** [zephyrproject-rtos/zephyr#60781] [Q&A] [answered]
Hi. First, some context:
We (Nordic) have our zephyr-based sdk https://github.com/nrfconnect/sdk-nrf. We want to improve our CI by introduction of variable testing scopes based on detected changes (less resources used, faster execution time). I was analyzing a workflow which runs in the upstream on PRs https://github.com/zephyrproject-rtos/zephyr/blob/main/.github/workflows/twister.yaml The workflow utilizes https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/ci/test_plan.py script to prepare a testplan.json an then twister loads it to get a scope for testing. I started to work on a changes that would be needed to allow to port such workflow to other repository, however, I have some questions regarding the workflow and the script.

1.) The script is called twice in the workflow https://github.com/zephyrproject-rtos/zephyr/blob/main/.github/workflows/twister.yaml#L87 and https://github.com/zephyrproject-rtos/zephyr/blob/main/.github/workflows/twister.yaml#L234 I understand, that the first call is to decide what is the scope and how many agents are to be started, but why the second call? Why not reusing the product of the first call?

2.) The script is called with --pull-request arg, however it looks like a dead variable which is not used in the script. Is there some idea behind it? 

2.a) Can --pull-request arg be used to turn on/off --integration mode? `find_tests(self)` is one of scope selection methods. If a test/sample yaml is modified, it is added to the scope. `--integration` mode in twister was introduced to define platforms on a test level for an on-PR scope. However, this option is not used in `find_tests`. When no specific platform was changed in a PR an option `--all` is used to get the scope resulting in all boards available in the usptream being added to the scope. Will it be beneficial to use `--integration` instead of `-all` when running on a pull request (--pull-request is used)? Will such scope be sufficient?

cc @nashif @stephanosio

**Comment by nashif** [comment, marked as answer]
> Can --pull-request arg be used to turn on/off --integration mode? `find_tests(self)` is one of scope selection methods. If a test/sample yaml is modified, it is added to the scope. `--integration` mode in twister was introduced to define platforms on a test level for an on-PR scope. However, this option is not used in `find_tests`. When no specific platform was changed in a PR an option `--all` is used to get the scope resulting in all boards available in the usptream being added to the scope. Will it be beneficial to use `--integration` instead of `-all` when running on a pull request (--pull-request is used)? Will such scope be sufficient?

in PRs, integration mode is the default, so if we want to end up using --integration, then basically this wofkflow will not be needed. The idea here is that if a test/sample was changed, we want to increase the scope and do more than the default, which is integration.

> 2.) The script is called with --pull-request arg, however it looks like a dead variable which is not used in the script. Is there some idea behind it?

Yes, probabyl can be removed, need to look at the history and see why it is not needed anymore.
